File: scripts/utils/cleanup-claude-json.py
#!/usr/bin/env python3
"""
Cleanup ~/.claude.json - Remove stale projects, caches, and bloat.
Hook: Can be run manually or via SessionEnd hook.

Usage:
    python cleanup-claude-json.py [--dry-run] [--verbose]
"""

import json
import os
import sys
import shutil
from pathlib import Path
from datetime import datetime

# Configuration
MAX_FILE_SIZE_MB = 5  # Warn if file exceeds this size
BACKUP_BEFORE_CLEANUP = True

def get_claude_json_path() -> Path:
    """Get path to ~/.claude.json"""
    return Path.home() / ".claude.json"

def get_file_size_mb(path: Path) -> float:
    """Get file size in MB"""
    if path.exists():
        return path.stat().st_size / (1024 * 1024)
    return 0

def backup_file(path: Path) -> Path:
    """Create a backup of the file"""
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    backup_path = path.parent / f".claude.json.backup_{timestamp}"
    shutil.copy2(path, backup_path)
    return backup_path

def cleanup_stale_projects(data: dict, dry_run: bool = False, verbose: bool = False) -> int:
    """Remove projects whose directories no longer exist"""
    removed = 0
    projects = data.get("projects", {})
    stale_projects = []

    for project_path in list(projects.keys()):
        if not Path(project_path).exists():
            stale_projects.append(project_path)
            if not dry_run:
                del projects[project_path]
            removed += 1
            if verbose:
                print(f"  [STALE] {project_path}")

    return removed

def cleanup_caches(data: dict, dry_run: bool = False, verbose: bool = False) -> int:
    """Clear cached data that can be regenerated"""
    cleaned = 0

    # Clear statsig caches (feature flags, regenerated on startup)
    if "cachedStatsigGates" in data and data["cachedStatsigGates"]:
        if verbose:
            print(f"  [CACHE] cachedStatsigGates ({len(data['cachedStatsigGates'])} entries)")
        if not dry_run:
            data["cachedStatsigGates"] = {}
        cleaned += 1

    if "cachedDynamicConfigs" in data and data["cachedDynamicConfigs"]:
        if verbose:
            print(f"  [CACHE] cachedDynamicConfigs ({len(data['cachedDynamicConfigs'])} entries)")
        if not dry_run:
            data["cachedDynamicConfigs"] = {}
        cleaned += 1

    return cleaned

def cleanup_tips_history(data: dict, dry_run: bool = False, verbose: bool = False) -> int:
    """Reset tips history counters (they just track how many times tips were shown)"""
    if "tipsHistory" in data and data["tipsHistory"]:
        count = len(data["tipsHistory"])
        if verbose:
            print(f"  [TIPS] tipsHistory ({count} entries)")
        if not dry_run:
            data["tipsHistory"] = {}
        return 1
    return 0

def cleanup_project_stats(data: dict, dry_run: bool = False, verbose: bool = False) -> int:
    """Reset per-project statistics (cost, tokens, duration, etc.)"""
    cleaned = 0
    stats_keys = [
        "lastCost", "lastAPIDuration", "lastAPIDurationWithoutRetries",
        "lastToolDuration", "lastDuration", "lastLinesAdded", "lastLinesRemoved",
        "lastTotalInputTokens", "lastTotalOutputTokens",
        "lastTotalCacheCreationInputTokens", "lastTotalCacheReadInputTokens",
        "lastModelUsage", "lastTotalWebSearchRequests"
    ]

    projects = data.get("projects", {})
    for project_path, project_data in projects.items():
        for key in stats_keys:
            if key in project_data:
                if not dry_run:
                    del project_data[key]
                cleaned += 1

    if verbose and cleaned > 0:
        print(f"  [STATS] Cleared {cleaned} stat entries across {len(projects)} projects")

    return cleaned

def cleanup_base64_content(data: dict, dry_run: bool = False, verbose: bool = False) -> int:
    """Remove any base64 encoded content (images, etc.)"""
    cleaned = 0

    def scan_and_clean(obj, path=""):
        nonlocal cleaned
        if isinstance(obj, dict):
            keys_to_delete = []
            for key, value in obj.items():
                if isinstance(value, str) and len(value) > 1000:
                    # Check if it looks like base64
                    if value.startswith("data:image") or (
                        len(value) > 5000 and
                        all(c in "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=" for c in value[:100])
                    ):
                        if verbose:
                            print(f"  [BASE64] {path}.{key} ({len(value)} chars)")
                        keys_to_delete.append(key)
                        cleaned += 1
                else:
                    scan_and_clean(value, f"{path}.{key}")

            if not dry_run:
                for key in keys_to_delete:
                    del obj[key]
        elif isinstance(obj, list):
            for i, item in enumerate(obj):
                scan_and_clean(item, f"{path}[{i}]")

    scan_and_clean(data)
    return cleaned

def main():
    dry_run = "--dry-run" in sys.argv
    verbose = "--verbose" in sys.argv or "-v" in sys.argv

    claude_json = get_claude_json_path()

    if not claude_json.exists():
        print(f"File not found: {claude_json}")
        return 1

    # Check file size
    size_mb = get_file_size_mb(claude_json)
    print(f"File: {claude_json}")
    print(f"Size: {size_mb:.2f} MB")

    if size_mb > MAX_FILE_SIZE_MB:
        print(f"WARNING: File exceeds {MAX_FILE_SIZE_MB} MB threshold!")

    if dry_run:
        print("\n[DRY RUN - No changes will be made]\n")

    # Load data
    with open(claude_json, "r", encoding="utf-8") as f:
        data = json.load(f)

    # Backup
    if BACKUP_BEFORE_CLEANUP and not dry_run:
        backup_path = backup_file(claude_json)
        print(f"Backup: {backup_path}\n")

    # Run cleanups
    print("Cleanup operations:")

    stale = cleanup_stale_projects(data, dry_run, verbose)
    print(f"  - Stale projects removed: {stale}")

    caches = cleanup_caches(data, dry_run, verbose)
    print(f"  - Caches cleared: {caches}")

    tips = cleanup_tips_history(data, dry_run, verbose)
    print(f"  - Tips history reset: {'yes' if tips else 'no'}")

    stats = cleanup_project_stats(data, dry_run, verbose)
    print(f"  - Project stats cleared: {stats}")

    base64 = cleanup_base64_content(data, dry_run, verbose)
    print(f"  - Base64 content removed: {base64}")

    # Save
    if not dry_run:
        with open(claude_json, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2, ensure_ascii=False)

        new_size_mb = get_file_size_mb(claude_json)
        saved = size_mb - new_size_mb
        print(f"\nNew size: {new_size_mb:.2f} MB (saved {saved:.2f} MB)")
    else:
        print("\n[DRY RUN COMPLETE - Run without --dry-run to apply changes]")

    return 0

if __name__ == "__main__":
    sys.exit(main())

File: scripts/utils/cleanup-claude-json.js
#!/usr/bin/env node
/**
 * Cleanup ~/.claude.json - Remove stale projects, caches, and bloat.
 *
 * Can be run manually or via SessionEnd hook.
 * Usage: node cleanup-claude-json.js [--dry-run] [--verbose]
 *
 * Pattern from cleanup-claude-json.py
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// Configuration
const MAX_FILE_SIZE_MB = 5; // Warn if file exceeds this size
const BACKUP_BEFORE_CLEANUP = true;

/**
 * Get path to ~/.claude.json
 *
 * @returns {string} Path to claude.json
 */
function getClaudeJsonPath() {
    return path.join(os.homedir(), '.claude.json');
}

/**
 * Get file size in MB
 *
 * @param {string} filePath - Path to file
 * @returns {number} File size in MB
 */
function getFileSizeMB(filePath) {
    if (!fs.existsSync(filePath)) {
        return 0;
    }
    const stats = fs.statSync(filePath);
    return stats.size / (1024 * 1024);
}

/**
 * Create a timestamped backup of the file
 *
 * @param {string} filePath - Path to file to backup
 * @returns {string} Path to backup file
 */
function backupFile(filePath) {
    const now = new Date();
    const timestamp = now.toISOString()
        .replace(/[-:]/g, '')
        .replace('T', '_')
        .slice(0, 15); // YYYYMMDD_HHMMSS

    const backupPath = path.join(
        path.dirname(filePath),
        `.claude.json.backup_${timestamp}`
    );

    fs.copyFileSync(filePath, backupPath);
    return backupPath;
}

/**
 * Remove projects whose directories no longer exist
 *
 * @param {Object} data - Claude.json data
 * @param {boolean} dryRun - If true, don't modify data
 * @param {boolean} verbose - If true, print verbose output
 * @returns {number} Number of stale projects removed
 */
function cleanupStaleProjects(data, dryRun = false, verbose = false) {
    let removed = 0;
    const projects = data.projects || {};

    for (const projectPath of Object.keys(projects)) {
        if (!fs.existsSync(projectPath)) {
            if (verbose) {
                console.log(`  [STALE] ${projectPath}`);
            }

            if (!dryRun) {
                delete projects[projectPath];
            }

            removed++;
        }
    }

    return removed;
}

/**
 * Clear cached data that can be regenerated
 *
 * @param {Object} data - Claude.json data
 * @param {boolean} dryRun - If true, don't modify data
 * @param {boolean} verbose - If true, print verbose output
 * @returns {number} Number of caches cleared
 */
function cleanupCaches(data, dryRun = false, verbose = false) {
    let cleaned = 0;

    // Clear statsig caches (feature flags, regenerated on startup)
    if (data.cachedStatsigGates && Object.keys(data.cachedStatsigGates).length > 0) {
        if (verbose) {
            console.log(`  [CACHE] cachedStatsigGates (${Object.keys(data.cachedStatsigGates).length} entries)`);
        }

        if (!dryRun) {
            data.cachedStatsigGates = {};
        }

        cleaned++;
    }

    if (data.cachedDynamicConfigs && Object.keys(data.cachedDynamicConfigs).length > 0) {
        if (verbose) {
            console.log(`  [CACHE] cachedDynamicConfigs (${Object.keys(data.cachedDynamicConfigs).length} entries)`);
        }

        if (!dryRun) {
            data.cachedDynamicConfigs = {};
        }

        cleaned++;
    }

    return cleaned;
}

/**
 * Reset tips history counters
 *
 * @param {Object} data - Claude.json data
 * @param {boolean} dryRun - If true, don't modify data
 * @param {boolean} verbose - If true, print verbose output
 * @returns {number} 1 if tips history was reset, 0 otherwise
 */
function cleanupTipsHistory(data, dryRun = false, verbose = false) {
    if (data.tipsHistory && Object.keys(data.tipsHistory).length > 0) {
        const count = Object.keys(data.tipsHistory).length;

        if (verbose) {
            console.log(`  [TIPS] tipsHistory (${count} entries)`);
        }

        if (!dryRun) {
            data.tipsHistory = {};
        }

        return 1;
    }

    return 0;
}

/**
 * Reset per-project statistics
 *
 * @param {Object} data - Claude.json data
 * @param {boolean} dryRun - If true, don't modify data
 * @param {boolean} verbose - If true, print verbose output
 * @returns {number} Number of stat entries cleared
 */
function cleanupProjectStats(data, dryRun = false, verbose = false) {
    let cleaned = 0;

    const statsKeys = [
        'lastCost', 'lastAPIDuration', 'lastAPIDurationWithoutRetries',
        'lastToolDuration', 'lastDuration', 'lastLinesAdded', 'lastLinesRemoved',
        'lastTotalInputTokens', 'lastTotalOutputTokens',
        'lastTotalCacheCreationInputTokens', 'lastTotalCacheReadInputTokens',
        'lastModelUsage', 'lastTotalWebSearchRequests'
    ];

    const projects = data.projects || {};

    for (const projectPath in projects) {
        const projectData = projects[projectPath];

        for (const key of statsKeys) {
            if (key in projectData) {
                if (!dryRun) {
                    delete projectData[key];
                }

                cleaned++;
            }
        }
    }

    if (verbose && cleaned > 0) {
        console.log(`  [STATS] Cleared ${cleaned} stat entries across ${Object.keys(projects).length} projects`);
    }

    return cleaned;
}

/**
 * Remove any base64 encoded content (images, etc.)
 *
 * @param {Object} data - Claude.json data
 * @param {boolean} dryRun - If true, don't modify data
 * @param {boolean} verbose - If true, print verbose output
 * @returns {number} Number of base64 fields removed
 */
function cleanupBase64Content(data, dryRun = false, verbose = false) {
    let cleaned = 0;

    function scanAndClean(obj, pathStr = '') {
        if (typeof obj === 'object' && obj !== null && !Array.isArray(obj)) {
            const keysToDelete = [];

            for (const [key, value] of Object.entries(obj)) {
                if (typeof value === 'string' && value.length > 1000) {
                    // Check if it looks like base64
                    const isDataImage = value.startsWith('data:image');
                    const looksLikeBase64 = value.length > 5000 &&
                        value.slice(0, 100).split('').every(c =>
                            'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='.includes(c)
                        );

                    if (isDataImage || looksLikeBase64) {
                        if (verbose) {
                            console.log(`  [BASE64] ${pathStr}.${key} (${value.length} chars)`);
                        }

                        keysToDelete.push(key);
                        cleaned++;
                    }
                } else if (typeof value === 'object' && value !== null) {
                    scanAndClean(value, `${pathStr}.${key}`);
                }
            }

            if (!dryRun) {
                for (const key of keysToDelete) {
                    delete obj[key];
                }
            }
        } else if (Array.isArray(obj)) {
            obj.forEach((item, index) => {
                scanAndClean(item, `${pathStr}[${index}]`);
            });
        }
    }

    scanAndClean(data);
    return cleaned;
}

/**
 * Main execution function
 *
 * @returns {Promise<number>} Exit code
 */
async function main() {
    const dryRun = process.argv.includes('--dry-run');
    const verbose = process.argv.includes('--verbose') || process.argv.includes('-v');

    const claudeJson = getClaudeJsonPath();

    if (!fs.existsSync(claudeJson)) {
        console.log(`File not found: ${claudeJson}`);
        return 1;
    }

    // Check file size
    const sizeMB = getFileSizeMB(claudeJson);
    console.log(`File: ${claudeJson}`);
    console.log(`Size: ${sizeMB.toFixed(2)} MB`);

    if (sizeMB > MAX_FILE_SIZE_MB) {
        console.log(`WARNING: File exceeds ${MAX_FILE_SIZE_MB} MB threshold!`);
    }

    if (dryRun) {
        console.log('\n[DRY RUN - No changes will be made]\n');
    }

    // Load data
    const content = fs.readFileSync(claudeJson, 'utf8');
    const data = JSON.parse(content);

    // Backup
    if (BACKUP_BEFORE_CLEANUP && !dryRun) {
        const backupPath = backupFile(claudeJson);
        console.log(`Backup: ${backupPath}\n`);
    }

    // Run cleanups
    console.log('Cleanup operations:');

    const stale = cleanupStaleProjects(data, dryRun, verbose);
    console.log(`  - Stale projects removed: ${stale}`);

    const caches = cleanupCaches(data, dryRun, verbose);
    console.log(`  - Caches cleared: ${caches}`);

    const tips = cleanupTipsHistory(data, dryRun, verbose);
    console.log(`  - Tips history reset: ${tips ? 'yes' : 'no'}`);

    const stats = cleanupProjectStats(data, dryRun, verbose);
    console.log(`  - Project stats cleared: ${stats}`);

    const base64 = cleanupBase64Content(data, dryRun, verbose);
    console.log(`  - Base64 content removed: ${base64}`);

    // Save
    if (!dryRun) {
        fs.writeFileSync(claudeJson, JSON.stringify(data, null, 2), 'utf8');

        const newSizeMB = getFileSizeMB(claudeJson);
        const saved = sizeMB - newSizeMB;
        console.log(`\nNew size: ${newSizeMB.toFixed(2)} MB (saved ${saved.toFixed(2)} MB)`);
    } else {
        console.log('\n[DRY RUN COMPLETE - Run without --dry-run to apply changes]');
    }

    return 0;
}

// CLI wrapper
/* istanbul ignore next */
if (require.main === module) {
    main().then(exitCode => {
        process.exit(exitCode);
    }).catch(error => {
        console.error('Error:', error.message);
        process.exit(1);
    });
}

module.exports = {
    getClaudeJsonPath,
    getFileSizeMB,
    backupFile,
    cleanupStaleProjects,
    cleanupCaches,
    cleanupTipsHistory,
    cleanupProjectStats,
    cleanupBase64Content,
    main
};
